--- a/fs/aufs/i_op.c	2016-05-31 19:00:30.134549000 +0200
+++ b/fs/aufs/i_op.c	2016-05-31 19:42:46.808881736 +0200
@@ -577,14 +577,14 @@ out:
 void au_pin_hdir_set_owner(struct au_pin *p, struct task_struct *task)
 {
 #if defined(CONFIG_DEBUG_MUTEXES) || defined(CONFIG_SMP)
-	p->hdir->hi_inode->i_mutex.owner = task;
+	p->hdir->hi_inode->i_rwsem.owner = task;
 #endif
 }
 
 void au_pin_hdir_acquire_nest(struct au_pin *p)
 {
 	if (p->hdir) {
-		mutex_acquire_nest(&p->hdir->hi_inode->i_mutex.dep_map,
+		rwsem_acquire_nest(&p->hdir->hi_inode->i_rwsem.dep_map,
 				   p->lsc_hi, 0, NULL, _RET_IP_);
 		au_pin_hdir_set_owner(p, current);
 	}
@@ -594,7 +594,7 @@ void au_pin_hdir_release(struct au_pin *
 {
 	if (p->hdir) {
 		au_pin_hdir_set_owner(p, p->task);
-		mutex_release(&p->hdir->hi_inode->i_mutex.dep_map, 1, _RET_IP_);
+		rwsem_release(&p->hdir->hi_inode->i_rwsem.dep_map, 1, _RET_IP_);
 	}
 }
 
