--- a/fs/aufs/file.c	2016-05-31 19:02:45.842364000 +0200
+++ b/fs/aufs/file.c	2016-05-31 19:20:27.872523162 +0200
@@ -771,8 +771,7 @@ static int aufs_readpage(struct file *fi
 }
 
 /* it will never be called, but necessary to support O_DIRECT */
-static ssize_t aufs_direct_IO(struct kiocb *iocb, struct iov_iter *iter,
-			      loff_t offset)
+static ssize_t aufs_direct_IO(struct kiocb *iocb, struct iov_iter *iter)
 { BUG(); return 0; }
 
 /* they will never be called. */
