--- /dev/null
+++ b/block/bfq-cgroup.c
@@ -721,12 +721,6 @@ static struct cftype bfqio_files[] = {
 	},
 };
 
-static int bfqio_populate(struct cgroup_subsys *subsys, struct cgroup *cgroup)
-{
-	return cgroup_add_files(cgroup, subsys, bfqio_files,
-				ARRAY_SIZE(bfqio_files));
-}
-
 static struct cgroup_subsys_state *bfqio_create(struct cgroup *cgroup)
 {
 	struct bfqio_cgroup *bgrp;
@@ -837,8 +831,8 @@ struct cgroup_subsys bfqio_subsys = {
 	.can_attach = bfqio_can_attach,
 	.attach = bfqio_attach,
 	.destroy = bfqio_destroy,
-	.populate = bfqio_populate,
 	.subsys_id = bfqio_subsys_id,
+	.base_cftypes = bfqio_files,
 };
 #else
 static inline void bfq_init_entity(struct bfq_entity *entity,
